delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/service/DeliveryStatusUpdater.java

```diff
@@ -0,0 +1,59 @@
+package com.tracking_delivery_system.delivery_status_service.service;
+
+import com.tracking_delivery_system.delivery_status_service.model.Deliver;
+import com.tracking_delivery_system.delivery_status_service.model.DeliveryStatus;
+import com.tracking_delivery_system.delivery_status_service.model.StatusUpdate;
+import com.tracking_delivery_system.delivery_status_service.repository.DeliverRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.function.Predicate;
+
+@Service
+@RequiredArgsConstructor
+public class DeliveryStatusUpdater {
+    private final KafkaProducer kafkaProducer;
+    private final DeliverRepository repository;
+
+    private static final double AWAITING_PICKUP_PERCENTAGE = 1;
+    private static final double PICKED_UP_PERCENTAGE =  0.90;
+    private static final double IN_TRANSIT_PERCENTAGE = 0.50;
+    private static final double OUT_FOR_DELIVERY_PERCENTAGE = 0.10;
+    private static final double DELIVERED_PERCENTAGE = 0;
+
+    public void updateStatusBasedDistance(Long trackingProgress, Deliver deliver){
+        long totalDistance = deliver.getDistance();
+        UUID id = deliver.getId();
+        double percentageRemaining = (double) trackingProgress / totalDistance;
+
+        DeliveryStatus currentStatus = deliver.getStatus();
+        DeliveryStatus newStatus = getNewStatusBasedOnPercentage(percentageRemaining, currentStatus);
+
+        if (newStatus != currentStatus){
+            updateDeliveryStatus(deliver, newStatus, id);
+        }
+    }
+
+    private void updateDeliveryStatus(Deliver deliver, DeliveryStatus newStatus, UUID id) {
+        deliver.setStatus(newStatus);
+        StatusUpdate status = new StatusUpdate(deliver.getId(), newStatus.name());
+        kafkaProducer.sendUpdate("delivery-status", status);
+        repository.save(deliver);
+    }
+
+    private DeliveryStatus getNewStatusBasedOnPercentage(double percentageRemaining, DeliveryStatus currentStatus) {
+        Map<Predicate<Double>, DeliveryStatus> statusMap = new LinkedHashMap<>(){{
```

Comment:
this Map must be generated only one in constructor to not generate it every time we call getNewStatusBasedOnPercentage()